### PR TITLE
User creates tags via prediction description

### LINKED ISSUE #204: Be able to tag predictions in the initial prediction textfield
Currently, to give a tag to a prediction, you have to make a response prediction and enter a hashtag phrase (i.e. "#covid", "#politics", "#personal")

It'd be easier to explain tagging to new people, and easier to tag things myself, if I could tag predictions in the initial prediction textfield. 

A potential problem is that this might result in cluttered prediction text that's a bit harder to read, and redundant with the actual tags, so it might be good if the hashtags got trimmed out of the text.

### PR DESCRIPTION
**Pull request by jaysonvirissimo**
Closes #204 
prediction

### CODE REVIEW

File: app/methods/tag_adder.rb [hunk 1]
```diff
@@ -1,24 +1,39 @@
...
+    @save = save
+    @string = string || ''
   end
 
   def call
     if string.present? && prediction.present?
       tag_names.each do |tag_name|
 	      prediction.tag_names << tag_name
-      end && prediction.save
+      end && save && prediction.save
```
Comment (michaelwebb76) [inline app/methods/tag_adder.rb:12]: I find these "buried side effects" really cringeworthy now I work in Haskell most of the time. Would it be too much to ask that this class just modifies the prediction without persisting it, and it's up to the caller to decide whether or not to persist the modifications?

**Review by michaelwebb76 — COMMENTED** [review]
This is cool. From a UX perspective, I don't like that the tags "disappear" when I save my prediction. Could we add a display of each prediction's tags on the prediction show page so that the user doesn't wonder what happened to their tags? (The tags could also link to the "predictions with this tag" index page).

**Comment by jaysonvirissimo** [comment]
@michaelwebb76, currently the tags _are_ displayed on the prediction show page (beneath the prediction description) and _do_ link to the predictions index filtered by that tag, but that you didn't see them suggests they aren't obvious/are difficult to see, so perhaps we can make those stand out more.

File: app/methods/tag_adder.rb (hunk 1 above)
Comment (jaysonvirissimo) [reply app/methods/tag_adder.rb:12]: Agreed. Let me rework this.

**Comment by michaelwebb76** [comment]
@jaysonvirissimo Ah yes I see. Do you think it makes more sense _above_ the "Created by" line? IMHO I more expect to see it there, right next to the prediction description.

File: app/methods/tag_adder.rb (hunk 1 above)
Comment (jaysonvirissimo) [reply app/methods/tag_adder.rb]: [This](https://github.com/bellroy/predictionbook/pull/206/commits/9c1b284ae9d69c7d2b1af53138fce6ffb15299fa) commit gets `TagAdder` out of the prediction persistence business and also eliminates the need for the Boolean parameter.

**Comment by jaysonvirissimo** [comment]
@michaelwebb76, [this](https://github.com/bellroy/predictionbook/pull/206/commits/a2d9cf282460986f3bb72805ac3cd08d88a4e14b) moves the tags directly under the prediction description:

Screen Shot 2022-02-14 at 9 41 15 PM

Perhaps in the future we could make them stand out even more by styling them like labels.

**Comment by michaelwebb76** [comment]
Something like https://www.designlabthemes.com/css-tags-how-to-style-post-tags/

File: app/observers/response_observer.rb [hunk 2]
```diff
@@ -4,7 +4,13 @@ class ResponseObserver < ActiveRecord::Observer
...
-    response.prediction.response_notifications.each(&:new_activity!)
-    TagAdder.new(prediction: response.prediction, string: response.comment).call
+    response.prediction.tap do |prediction|
+      prediction.response_notifications.each(&:new_activity!)
+
+      tag_adder = TagAdder.new(prediction: response.prediction, string: response.comment)
+      if tag_adder.call
+        prediction.save
+      end
+    end
```
Comment (michaelwebb76) [inline app/observers/response_observer.rb:14]: This is heaps better.

File: app/views/predictions/show.html.erb [hunk 3]
```diff
@@ -20,6 +13,13 @@
   </p>
 <%end %>
 
+<p>
+  Created by <%= show_user(@prediction.creator) %>
+  <%= TimeInContentTagPresenter.new(@prediction.created_at).tag %>;
+  known <%= TimeInContentTagPresenter.new(@prediction.deadline).tag %><% if @prediction.judgement %>;
+  <%= render @prediction.judgement %><% end %>
+</p>
+
```
Comment (michaelwebb76) [inline app/views/predictions/show.html.erb:22]: 👌

**Review by michaelwebb76 — APPROVED** [review]

**Pull request merged**